refactor the transform model to be a bit more friendly with dirty state

bryanaka opened the pull request:
Refactors the Enum and subsequential transform paradigm to be a bit more friendly with DS Model dirty state in ember-data. 

### Changes
- A new API for rendering the Enum value 
- A new API for setting Enum value in Ember Data
- More efficiently creates the boolean computed properties i.e. `isStarted`
- Caches definitions created on the fly in an Enum Registry
- Refines the Enum data type 

### New Features
- Exposes `EnumRegistry` used for caching Enum definitions and registering custom definitions
- Exposes an `enumAttr` computed macro that allows a similar API as the Transform
- Exposes an API for controlling rendering in templates and serializing to server.

Sorry - I know this is a large change to the code base but this has been a burning need for me. When I stumbled across your library, I felt it was better to contribute to an existing lib than start a new lib 😅

Plus you got the good lib name already 😉  Please do not hesitate to reject this if you think it is the wrong direction for the library or suggest changes.

TODO
- [ ] Needs Tests
- [x] Change README to affect new API
- [x] More Documentation
- [ ] Feedback since this is a huge change to the library

bryanaka committed:
refactor the transform model to be a bit more freindly with dirty state in ember data. Also introduces a new API and Enum Type class definition paradigm

bryanaka committed:
clean up lint and some syntax errors (opps)

bryanaka committed:
alterations made to make slightly more performant - mutate the same enum type for the records so the cost of creating enums each time is smaller. change utils to be compatible with changed api. Added docs

bryanaka committed:
accidently added yarn lockfile

bryanaka committed:
Small changes to guard against incorrect values

- check if server return value is a valid option before setting.
- notify after setting value to correctly trigger observers.

bryanaka committed:
change README

bryanaka committed:
Fixing lint and naming issue

panthony commented on addon/transforms/enum.js:
```diff
@@ -1,22 +1,62 @@
-    } else {
-      return deserialized;
-    }
+    return get(deserialized, 'value') || deserialized.toString() || deserialized;
```
`Ember.get` shouldn't be called with `null` or `undefined` so maybe it shouldn't be called without any test on `deserialized` ?

Is there a case where `deserialized.toString()` is falsy and have a different value than `deserialized`?

bryanaka replied:
Don't think `deserialized` could be undefined since mutation just sets the value. *Maybe* if the user calls `delete`? seems like a huge anti-pattern though.

ah, good catch. we probably don't want `toString()` since we want that to be the handlebars rendering hook IMO.

panthony commented on addon/utils/create-enum-type.js:
```diff
@@ -0,0 +1,94 @@
+  return ENUM_TYPE_MAP[cacheKey];
+}
+
+export { createEnumType, ENUM_TYPE_MAP };
```
Not sure it's a good idea to export the var used as a cache holder, what's the purpose?

bryanaka replied:
Actually, I agree. The original idea was to allow others to register their own Enum types; maybe they want `toString` to return a humanized version of the value. This is a quick and dirty way of allowing it.

Ideally, I want to actually allow devs to register enum definitions in the Ember container, but this is a much bigger undertaking because now we got to loop in the resolver as well. Don't want to go down that path until the owner feels like this is a good direction.

I refactored a bit to use a registry object so the path forward to letting the container manage it is clear. Will be pushing it in a bit.

Thanks for the feedback!

panthony commented on README.md:
````diff
@@ -45,10 +95,21 @@ You can also use boolean methods to check if the enum has a certain value:
 
 ```handlebars
-<select value={{model.status.value}} onchange={{action (mut model.status.value) value="target.value"}}>
+<select value={{model.status}} onchange={{action (mut model.status) value="target.value"}}>
````
👍

panthony reviewed (commented):
I like this 👍

bryanaka committed:
Switch to using the registry object paradigm and create computed boolean definitions at extend time
- Enum can be safely extended without any additional work to manage boolean attributes.

bryanaka committed:
add name to readme

rmachielse commented:
@bryanaka thanks for this pr! I'm glad you thought of a way to make this more useful. It is indeed quite a large change. I will try to review it asap. Unfortunately I haven't got a lot of time currently. If you feel like maintaining the project I can transfer the ownership on npm to you so you can use your fork if you prefer.

bryanaka commented:
@rmachielse I would definitely be happy to maintain it! I have some free time right now and could use something to hack on.

If you do decide to transfer it to me, I can also add you as a contributor or co-owner (is that a thing?). No pressure to work on it, just want to make sure you have access to work on it should you ever decide to.

bryanaka committed:
Clean up code

bryanaka committed:
remove create-enum-type and use new registry api instead

This is made to mimic the factoryFor methods used in Ember’s container. This will make it easier for someone with Ember experience to pick up on the patters and/or map this to the container api later.

bryanaka commented:
Pretty happy with the API at this point. Over the next few days, I'll be working on getting the tests in shape.

Let me know if either of you see something you totally disagree with and we can discuss changes.

bryanaka committed:
docs

panthony commented:
Looks ok to me although I read rather rapidly. I'll try this PR this week or next week.

bryanaka commented:
btw I am `@bryanaka` on NPM as well.

panthony commented:
Quick question @bryanaka, how do you import the enums you directly add to the registry ? May be a dumb question but I'm not sure how to bootstrap this in my app.

I started with a directory such as:

`/app/enums/:name.js`



I need to do it like this because:
- I use the same enums in different models (currencies, intervals, ...)
- I need to add a humanized version of the enum (and I do not want to repeat it elsewhere)
- I need to provide the enum options (and humanized values) to a components (radio, select, ...) that are not related to a given model.

bryanaka commented:
@panthony yeah, you have to manually call `EnumRegistry#register` on the class. Autoloading and registering isn't a thing yet - was going to tackle after I get the tests passing.

I actually think I found a bug in the implementation for dirty still, so may want to hold off. I will ping you once I push what I consider a workable beta

bryanaka committed:
no message

rmachielse commented:
@bryanaka why do you use a custom registry? I feel this change is an overengineering and the original problem can be solved much simpler.

bryanaka commented:
@rmachielse Yeah. I am still looking for a way around that. The problem is this-

`DS.attr` returns a computed property with a getter and setter. The getter and setter essentially are aliases to the state on the internal model or an alias to the defaultValue (which gets called if a function). 

On `serialize` and `deserialize` you can return an `Enum` and you will be fine, but this means if you set the property on the client side, you would also need to create an enum and set it to that to trigger proper dirty state. So there is a lot of ceremony around that way.

My original solution was having the serializer and defaultValue return an `Enum` wrapped in a computed property to intercept `sets` but this doesn't work because the computed property that `DS.attr` returns does not expect/work with the value being a computed property instance.

So at the end of the day, I think we are back to having to rely on the end-user setting with an Enum instance unless we used this super hacky way (which I agree is not ideal) of getting around it.

bryanaka commented:
Putting this PR on hold temporarily while I ping-pong hack on this with @runspired 

If either of you are interested in participating in the convo, here is the beginning of the tweet chain

https://twitter.com/BryanAKA/status/826169371185668096

rmachielse commented:
@bryanaka how are you doing on this, did you get any further on that discussion? Did you solve your problem in a different way?

bryanaka commented:
@rmachielse not really, it seems there is essentially no non-hacky way to do it at this time.

It can be accomplished by reaching into the caching internals and/or the ComputedProperty private API but not sure if that's a trade up here.

Either way, there are better ways than this PR. Closing this and may open a new one when I get some time. Maybe @runspired has thoughts from when we were going back and forth on this.
